### ISSUE TITLE
Tests with `Isolate.run` pause indefinitely.

### ISSUE THREAD

**Issue by HosseinYousefi** [dart-lang/coverage#472]
This never prints 42 when it's run using `--pause-isolates-on-exit`.

```
import 'dart:isolate';

void main() async {
  print(await Isolate.run(() => 42));
}
```

And since coverage uses the `--pause-isolates-on-exit` (presumably to communicate with the VM service), the tests time out.

**Comment by lrhn** [comment]
The `Isolate.run` implementation runs `Isolate.exit(resultMessage, responsePort);` when it's done.
That should shut down the isolate (prevent any further code executing) and send a message on `responsePort`
*before* sending any `onExitListener` messages (that will send a `null` on the same port, which will have been closed when the result message arrived).

If sending the `resultMessage` gets blocked on `--pause-isolates-on-exit`, before sending the result,
the the original isolate won't receive the message, and the `await` won't complete.

In that case, it's pretty much working as intended.

And it pretty much *has* to work that way, because the `resultMessage` isn't as much *sent* as its heap is merged into the receiver's heap, at which point it no longer belongs to the exiting isolate. If paused after that, it would be impossible to do anything with the paused isolate anyway, since its heap is no longer its own.

So "pause-isolates-on-exit" probably _has_ to mean "pause-before-merging-heaps", which means it happens before, and blocks, the result message being received.

**Comment by dcharkes** [comment]
Thanks @lrhn! That's what @HosseinYousefi and I thought as well, thanks for confirming!

So the question is how that would work together with coverage, do all the isolates need to stay in a paused state to get full coverage info, or do we just need the main isolate to pause on exit? E.g. would a hypothetical `--pause-main-isolate-on-exit` suffice for coverage collection?

**Comment by cbracken** [comment]
@lrhn covered it above, yep. `--pause-isolates-on-exit` was indeed added so that we could keep all isolates around in a wedged state once they've wrapped up, then collect coverage from them before finally unwedging and bailing out.

The knock-on effect of the stuck `resultMessage` wasn't something I considered.

I can't remember if we currently wait for _all_ isolates to be paused before collecting coverage, but that sounds plausible. Perhaps we could collect coverage on each isolate once we determine it's paused, then unpause them once we've collected, in an iterative fashion until we're down to the last one? Apologies, it's been years since I've looked at this code.

**Comment by dcharkes** [comment]
> Perhaps we could collect coverage on each isolate once we determine it's paused, then unpause them once we've collected, in an iterative fashion until we're down to the last one?

@bkonyi You probably know if this is easy to do with the dev protocol.

**Comment by bkonyi** [comment]
Yep, this isn't hard to do if each isolate was configured to pause on exit (simply providing `--pause-isolates-on-exit` will accomplish that).

First, you'll want to listen to the VM service's `Isolate` stream to get isolate events:

```dart
service.onIsolateEvent.listen((event) {
  if (event.kind == EventKind.kIsolateStart) {
    // An isolate has started
  } else if (event.kind == EventKind.kIsolateExit) {
    // An isolate has exited
  }
});
await service.streamListen(EventStreams.kIsolate);
```

Then, since there's a chance you might have missed an isolate pause event while listening to the isolate stream, you'll want to check if any isolates are already paused:

```dart
// Get the VM state
final vm = await service.getVM();

for (final isolateRef in vm.isolates!) {
  // Get the isolate's current state
  final isolate = await service.getIsolate(isolateRef.id!);
  if (isolate.pauseEvent?.kind == EventKind.kIsolateExit) {
     // isolate has paused at exit.
  }
}
```

You'll need to keep track of how many isolates have been spawned and how many have exited, but then it should be easy to collect coverage and resume each isolate to cause the process to exit.
